crates/core/src/transport/udp_connection.rs

```diff
@@ -0,0 +1,187 @@
+use crate::transport::udp_transport::UdpTransport;
+use crate::transport::ConnectionEvent;
+use aes_gcm::{
+    aead::{generic_array::GenericArray, Aead},
+    Aes128Gcm,
+};
+use libp2p_identity::PublicKey;
+use rand::rngs::OsRng;
+use rand::RngCore;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::{mpsc, Mutex, RwLock};
+use tokio::task;
+
+/*
+ NOTES:
+    The receiver thread should be set up when the channel is created, it shouldn't be stored
+    in the struct because only the receiver thread loop should be able to access it.
+
+
+*/
+
+pub struct UdpConnection {
+    transport: Arc<RwLock<UdpTransport>>,
```

Comment:
this is a bit of a smell to me, I am not sure why we need any of this to be multithreaded and hide it behing a lock (!)

Connections should be handled ina  single thread and if you need to open multiple then you handle a pool of connections in multiple threads, but all the code in this PR feels to me should be single-threaded.